consensus/core/src/transaction.rs

```diff
@@ -145,24 +191,29 @@ impl TransactionClient {
     /// has not been included to any block.
     pub(crate) async fn submit_no_wait(
         &self,
-        transaction: Vec<u8>,
+        transactions: Vec<Vec<u8>>,
     ) -> Result<oneshot::Receiver<()>, ClientError> {
         let (included_in_block_ack_send, included_in_block_ack_receive) = oneshot::channel();
-        if transaction.len() as u64 > self.max_transaction_size {
-            return Err(ClientError::OversizedTransaction(
-                transaction.len() as u64,
-                self.max_transaction_size,
-            ));
+        for transaction in &transactions {
+            if transaction.len() as u64 > self.max_transaction_size {
+                return Err(ClientError::OversizedTransaction(
+                    transaction.len() as u64,
+                    self.max_transaction_size,
+                ));
+            }
         }
 
         let t = TransactionGuard {
-            transaction: Transaction::new(transaction),
+            transactions: transactions
+                .into_iter()
+                .map(|x| Transaction::new(x))
+                .collect(),
             included_in_block_ack: included_in_block_ack_send,
         };
         self.sender
```

Comment:
Thanks for pointing this out!

I guess in that case we need to spawn a separated task to wait on multiple oneshot channels (one for each TransactionGuard), then after all channels have been signalled, the task eventually signals the oneshot channel returned by `submit_no_wait`.  I am not quite familiar with the runtime so not sure how much burden will this create.

Or we could actually change the API to remove the 'no wait' semantic - since it is already async.

I can proceed either way if people are okay with it.

Reply:
I think we can use something like `futures::future::join_all()` to await on all oneshot channels without spawning a new task. But I'm not sure if it is an improvement over using only one ack callback. Maybe keep the current logic and we can iterate on this part in future? Wdyt @akichidis?

Reply:
Yeah anything like `futures::future::join_all()` would do here , even a simple loop, as it's all or nothing. @mwtian it's not a performance improvement but rather , to my view, a complexity improvement. I felt that with the current approach there is lots for break up/ management in the `TransactionConsumer` which I would avoid. Just to be clear, not a deal breaker as solution, but that would be my preference.